### PR TITLE
TIP-451: split product normalizer into properties and associations normalizers

### PR DESCRIPTION
**Pull request by jjanvier**
The idea here is that the product normalizer always returns the "standard" format (ie: it always returns a product with *all its data*). No more weird options like _exclude_associations_ or _only_associations_.

But in really particular cases, for performances reasons only, we need to normalize only the associations or the properties of the products. That's why those 2 normalizers have been introduced. And the product normalizer now uses them.

The `pim_webservice_serializer` has been removed as it was totally useless. It was the same than the `pim_serializer`.

Also the get product REST API has been fixed. The filtering of the product values was not correct in EE :(

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Changelog updated                 | Y
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductAssociationsNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     * {@inheritdoc}
+     *
+     * @param ProductInterface $product
+     */
+    public function normalize($product, $format = null, array $context = [])
+    {
+        $data = [];
+
+        foreach ($product->getAssociations() as $association) {
+            $code = $association->getAssociationType()->getCode();
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductAssociationsNormalizer.php]: seen IRL, initialize `$data[$code]['groups']` and `$data[$code]['products']` please.

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductNormalizer.php [hunk 2]
```diff
@@ -2,99 +2,48 @@
...
 
+    /** @var NormalizerInterface */
+    protected $propertiesNormalizer;
+
+    /** @var NormalizerInterface */
+    protected $associationsNormalizer;
+
     /**
-     * @param CollectionFilterInterface $filter The collection filter
+     * ProductNormalizer constructor.
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductNormalizer.php]: please configure your IDE to do not generate automatically this kind code ;)

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+use Pim\Bundle\CatalogBundle\Filter\CollectionFilterInterface;
+use Pim\Component\Catalog\Model\ProductInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\SerializerAwareNormalizer;
+
+/**
+ * A normalizer to transform product properties into an array
+ *
+ * @author    Filips Alpe <filips@akeneo.com>
+ * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: Seems to be represented as a new file by Github, but maybe it's a move... in both cases, you can take it ;)

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,133 @@
+<?php
+
+namespace Pim\Component\Catalog\Normalizer\Structured;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Pim\Bundle\CatalogBundle\Filter\CollectionFilterInterface;
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: A reference to `Pim\Bundle` into a component ?

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        $context = array_merge($defaultContext, $context);
+        $context['entity'] = 'product';
+        $data = [];
+
+        if (isset($context['resource'])) {
+            $data['resource'] = $context['resource'];
+        }
+
+        $data[self::FIELD_FAMILY]        = $product->getFamily() ? $product->getFamily()->getCode() : null;
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: null === $product->getFamily (and the other ones too)

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php (hunk 4 above)
Comment (skeleton) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: seen IRL, will be remove with my PR

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    /**
+     * @param ProductInterface $product
+     *
+     * @return array
+     */
+    protected function getGroups(ProductInterface $product)
+    {
+        $groups = [];
+
+        if ($product->getGroupCodes()) {
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: You can drop this line

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            $groups = $product->getGroupCodes();
+            if ($product->getVariantGroup()) {
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: null !==

**Comment by pierallard** [comment]
Some boyscoutism about Filip's work, GTM anyway.

**Comment by skeleton** [comment]
GTM once fixed

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductAssociationsNormalizer.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductAssociationsNormalizer.php]: done

Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductAssociationsNormalizer.php]: I remark this thing too...
Does PHP auto assign the keys ? Is it just a "good practice" to declare it ? Real question.

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php (hunk 3 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: I copied/pasted the ProductNormalizer made by Filips
and removed all the "associations" related things
That's why I let its credits

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php (hunk 5 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: done

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php (hunk 6 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: done

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php (hunk 7 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPropertiesNormalizer.php]: done

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductAssociationsNormalizer.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductAssociationsNormalizer.php]: yes PHP auto assign keys
but yes, it's "better" to declare it :)

**Pull request merged**